flocker/ca/test/test_ca.py

```diff
@@ -0,0 +1,171 @@
+"""
+
+import datetime
+import os
+
+from Crypto.Util import asn1
+from OpenSSL import crypto
+
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.python.filepath import FilePath
+
+from .. import CertificateAuthority, PathError, EXPIRY_20_YEARS
+
+
+class CertificateAuthorityTests(SynchronousTestCase):
+    """
+    Tests for ``flocker.ca._ca.CertificateAuthority``.
+    """
+
+    def test_written_keypair_exists(self):
+        """
+        ``CertificateAuthority.initialize`` writes a PEM file to the
+        specified path.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        CertificateAuthority.initialize(path, b"mycluster")
+        self.assertEqual(
+            (True, True),
+            (path.child("cluster.crt").exists(),
+             path.child("cluster.key").exists())
+        )
+
+    def test_decoded_certificate_matches_public_key(self):
+        """
+        A decoded certificate's public key matches the public key it is
+        meant to be paired with.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        ca = CertificateAuthority.initialize(path, b"mycluster")
+        self.assertTrue(
+            ca.keypair.keypair.matches(ca.certificate.getPublicKey())
+        )
+
+    def test_decoded_certificate_matches_private_key(self):
+        """
+        A decoded certificate matches the private key it is meant to
+        be paired with.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        ca = CertificateAuthority.initialize(path, b"mycluster")
+        priv = ca.keypair.keypair.original
+        pub = ca.certificate.getPublicKey().original
+        pub_asn1 = crypto.dump_privatekey(crypto.FILETYPE_ASN1, pub)
+        priv_asn1 = crypto.dump_privatekey(crypto.FILETYPE_ASN1, priv)
+        pub_der = asn1.DerSequence()
+        pub_der.decode(pub_asn1)
+        priv_der = asn1.DerSequence()
+        priv_der.decode(priv_asn1)
+        pub_modulus = pub_der[1]
+        priv_modulus = priv_der[1]
+        self.assertEqual(pub_modulus, priv_modulus)
+
+    def test_written_keypair_reloads(self):
+        """
+        A keypair written by ``CertificateAuthority.initialize`` can be
+        successfully reloaded in to an identical ``CertificateAuthority``
+        instance.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        ca1 = CertificateAuthority.initialize(path, b"mycluster")
+        ca2 = CertificateAuthority.from_path(path)
+        self.assertEqual(ca1, ca2)
+
+    def test_keypair_correct_umask(self):
+        """
+        A keypair file written by ``CertificateAuthority.initialize`` has
+        the correct permissions (0600).
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        CertificateAuthority.initialize(path, b"mycluster")
+        keyPath = path.child(b"cluster.key")
+        st = os.stat(keyPath.path)
```

Comment:
Maybe it'd be better to use FilePath.getPermissions here:
 * http://twistedmatrix.com/documents/current/api/twisted.python.filepath.FilePath.html#getPermissions